### PR TITLE
Support multiple segments in-flight between B & O

### LINKED ISSUE #1658: More flexible segment in flight rule
Both B & O implement a single segment in flight (SSIF) rule for streams. B will not send another segment to O if it has not responded for a previous segment for the stream yet. O will not accept another segment if it has not finished transcoding a previous segment for a transcode session yet. As a result, for a stream with uniform segment durations, an O will receive all segments in the stream as long as it returns results in real-time. However, there are other considerations to take into account in practice:

- If a stream has variable segment durations (which is not uncommon), segment N + 1 might be shorter than segment N such that even if O can return segment N in real-time, segment N + 1 could be emitted before O returns segment N
- O is slower to transcode the first segment in a transcode session due to GPU session init overhead
- O might return results in real-time most of the time, but slower than real-time sometimes due to factors such as fluctuation in bandwidth capacity
- Combination of all of the above

The points listed above can cause the following problems:
- If B only has access to a single O, it will drop a bunch segments even if those segments could have been transcoded by O in real-time most of the time and an acceptable amount slower than real-time other times
- If B has access to many Os, it will switch Os more often and each switch can potentially trigger a GPU session init which will slow down the transcode for that segment

One way to address these problems might be to update the segment in flight rule to something like:

- If a segment is in flight for O:
    - If the time in flight for the segment is less than `x`:
        - Send the next segment to this O
    - Else:
        - If there are other Os available:
            - Send the next segment to another O
        - Else:
            - If the time in flight for the segment is less than `y`:
                - Send the next segment to this O
            - Else:
                - Drop the segment

The above assumes that the current O would be the one selected for the next segment by B's selection algo. 

Since this might result in multiple segments in flight, we could keep track of the time in flight for the oldest segment in flight.

`x` is the time in flight threshold for continuing to send segments to an O. A default value for this might be the duration of the oldest segment that is in flight. 

`y` is the time in flight threshold for continuing to send segments to an O if the requirement for `x` is not met and if there are other Os available. This value controls how often segments get dropped.

A few implications of this proposed rule if the suggested default value for `x` is used:
- Regardless of the durations of segments in a stream, if an O can return results for each segment in real-time it will receive all the segments of the stream.
- If O fluctuates between real-time and slightly slower than real-time, it can still consistently receive segments of the stream if there are no other Os available which reduces the # of dropped segments
- If there are other Os available, B will still switch if the current Os is not keeping up in real-time

This would require changes to the segment in flight management rule on both B & O.

### PR DESCRIPTION
**Pull request by jailuthra**
**What does this pull request do? Explain your changes. (required)**

Implements flexible segment-in-flight handling between a B and an O as described in https://github.com/livepeer/go-livepeer/issues/1658. Previously only a single segment could be in flight on a particular B<>O session.

**Specific updates (required)**

- https://github.com/livepeer/go-livepeer/pull/1735/commits/23a7532fc858f23e105a6922b4139fddfaf1dde5 - Add a new metric to count Orchestrator Swaps (needed by https://github.com/livepeer/docker-livepeer/pull/50)
- https://github.com/livepeer/go-livepeer/pull/1735/commits/5dc0b25f1f58043f8bc6a66072b2a6098c08941d 
    - Track the last used BroadcasterSession in a `lastSession` variable and also track all segments-in-flight for a particular session in a queue.
    - Re-use the last session whenever possible - as specified in https://github.com/livepeer/go-livepeer/issues/1658
    - Increment the new metric for "Orchestrator Swaps" whenever new session's O URI is different from last session's O URI

**How did you test each of these updates (required)**

A lot of manual testing -
1. Ensuring session re-use when simultaneous segments arriving at the same session (i.e. get a 200 response for all segments when sent as described in https://github.com/livepeer/go-livepeer/issues/1665 even when B knows only one O)
2. Ensuring Segments-In-Flight queue is cleaned up when all segments return for a particular session
3. Check the new `mOrchestratorSwaps` metric is updated whenever a swap occurs

**TODO**
- [x] Fix existing Unit Tests
- [x] Write unit tests for new code
- [x] Remove Log lines
- [x] Race condition checks

**Does this pull request close any open issues?**

Fixes #1658 and #1719

**Checklist:**

- [ ] README and other documentation updated
- [ ] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

File: monitor/census.go
Comment (yondonfu) [inline monitor/census.go]: Since `orchestratorSwapped()` is only called here, could we just move the code from that method here similar to what is done in `CurrentSessions()` below?

File: server/broadcast.go
Comment (yondonfu) [inline server/broadcast.go]: Moving this method call outside of `checkSessions()` means that `checkSessions()` will not see any updates to the number of sessions which is updated by the `refreshSessions()`. I think this can result in the following scenario:

- `numSess > 0`
- `checkSessions()` starts a session refresh by calling `refreshSessions()` because `numSess` is below the desired threshold
- `checkSessions()` returns true
- If the loop on L91 runs more than once (i.e. if the first selected session has already been removed from `sessMap`)
    - `checkSessions()` always returns true because `numSess` is never updated
    - `refreshSessions()` is always called in `checkSessions()` because `numSess` is still below the desired threshold

Can we preserve the currently property of `checkSessions()` always seeing updates to `numSess`?

File: server/broadcast.go
Comment (yondonfu) [inline server/broadcast.go]: Is the `numSess` check here necessary? Would it suffice to just use an else block here since `sel.Select()` would return nil if there are no sessions left and we have a nil check for the returned session on L105?

File: server/broadcast.go
Comment (yondonfu) [inline server/broadcast.go]: Possible race here since we read SegsInFlight without a lock? (Not a problem if this log is going to be removed anyway).

File: server/broadcast.go
Comment (yondonfu) [inline server/broadcast.go]: Could use `time.Duration(seg.Duration * float64(time.Second))` since I think seg.Duration is already a denominated in seconds to avoid extra unit conversions.

File: server/broadcast.go
Comment (yondonfu) [inline server/broadcast.go]: If we're returning in the if block, it's common to skip the else block in Go i.e.

```golang
if (len(bsm.lastSess.SegsInFlight) > 1 {
    ...
    return
}

...
bsm.lastSess = sess
```

File: server/broadcast.go
Comment (yondonfu) [inline server/broadcast.go]: If we're returning in the if block, it's common to skip the else if block in Go i.e.

```golang
if ... {
}

if bsm.lastSess != nil && ... {
}
```

**Review by yondonfu — CHANGES_REQUESTED** [review]
Took a first pass.

We should probably also make sure that the orchestrator can support multiple in-flight segments as mentioned [here](https://github.com/livepeer/go-livepeer/issues/1658#issuecomment-737423979). 

I think it'd be a good idea to run some of the unit tests for this new code with the race detector ([example](https://github.com/livepeer/go-livepeer/blob/897d5760c9aef0519f0dae0b4e34be7888a3a0bc/test.sh#L24)) since we had some concerns around concurrent session usage.

File: server/broadcast.go
Comment (jailuthra) [reply server/broadcast.go]: Ah I encountered an error when `sel.Select()` was called on an empty list with `LIFOSelector` but I guess a better solution would be to add a check in it's `Select()` method and return nil - rather than adding a check during calling. Will do in the update and revert back the `numSess` changes.

File: server/broadcast.go
Comment (jailuthra) [reply server/broadcast.go]: Yep planning to remove the log before push!

File: server/broadcast.go
Comment (jailuthra) [reply server/broadcast.go]: Good catch, I completely missed that we'd want to keep that variable updated between loop iterations.

**Comment by jailuthra** [comment]
- Fixed all review comments
- Completed Todo in PR description
- Added a new unit test to reproduce the manual testing done with multiple in-flight segments and verify the state variables are set correctly 
- Added race detection in `test.sh` script

Will squash/cleanup commit history before merge, once approved

File: core/orchestrator.go
Comment (yondonfu) [inline core/orchestrator.go]: Can we define a constant for this so its clear that its an important parameter that controls the maximum # of segments that can be queued on an orchestrators?

File: server/broadcast.go
Comment (yondonfu) [inline server/broadcast.go:136]: If the last session was removed from the map should we also set `bsm.lastSess = nil` here since we know that we wouldn't want to use the last session in the next iteration of the loop?

File: server/broadcast.go
Comment (yondonfu) [inline server/broadcast.go]: I think we should still set `bsm.lastSess` and update `sessMap` (L170) because the provided `sess` is going to be a new copy (see [this line](https://github.com/livepeer/go-livepeer/blob/897d5760c9aef0519f0dae0b4e34be7888a3a0bc/server/broadcast.go#L442)) that was updated based on an existing session. If we don't make these updates, we could end up using a session for the same orchestrator later on with out-dated metadata fields. We could write a test to verify this by changing the `OrchestratorInfo` message returned with a segment response.

File: server/broadcast_test.go
Comment (yondonfu) [inline server/broadcast_test.go:329]: I see that we're running tests prefixed with `TestSelectSession_` using the race detector, but I don't see any tests involving concurrency below. Concurrent access for the last session seems possible with:

- `selectSession()` called in `transcodeSegment()`
- `completeSession()` called in `transcodeSegment()`
- The update of the `SegsInFlight` for the last session in `transcodeSegment()`

Perhaps we can just setup scenarios where those 3 events occur concurrently (i.e. parallel goroutines).

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: server/broadcast.go
Comment (jailuthra) [reply server/broadcast.go:136]: This wasn't a hard requirement as we would also test for `len(bsm.lastSess.SegsInFlight) > 0` in next iteration and then end up using a new session or nil.
But adding this definitely makes the code more explicit/clear so done that in the update!

File: server/broadcast_test.go
Comment (jailuthra) [reply server/broadcast_test.go:329]: Will send another quick PR for this as this PR blocks other things!

**Review by yondonfu — APPROVED** [review]
LGTM after squashing

**Pull request merged**
